fsspec/core.py

```diff
@@ -432,6 +434,19 @@ def _fetch(self, start, end):
 
         return self.cache[start:end]
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        del state['cache']
+        del state['_file']
+        return state
+
+    def __setstate__(self, state):
+        # Restore instance attributes
+        self.__dict__.update(state)
+        self.blocks = set()
```

Comment:
Why not serialise this too? Blocks can only be added, so it will never go stale. Else, the data on disc is not actually useful for the reconstituted instance. I wonder if there's a way to tell which parts of a file contain any data...

Reply:
I actually just realized that intention is to persist the data on disk because `CachingFileSystem` does explicitly control the the underlying file. I think we should clear out blocks if the underlying file is a `TemporaryFile`, otherwise I agree with you that we can serialize `blocks` too.

Reply:
Probably we should be sure to update `self.location` to the created file, if it is temporary, and then we can safely reuse it and its blocks.

Reply:
Since the temporary file could be gone when this unpickled, so I do prefer clearing out the cache if `location` was not explicitly supplied. What do you think?

Reply:
I would say, check for the file's existence and if it's there, use it. Maybe we should store the fact that the original call was for a temporary file, so if the file is gone, we know to make a new one rather than error.